### PR TITLE
[SPARK-29850][SQL] sort-merge-join an empty table should not memory leak

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

When whole stage codegen `HashAggregateExec`, create the hash map when we begin to process inputs.

### Why are the changes needed?

Sort-merge join completes directly if the left side table is empty. If there is an aggregate in the right side, the aggregate will not be triggered at all, but its hash map is created during codegen and can't be released.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

a new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
note that, this is not a serious problem. The memory manager always releases all the allocated memory at the end of a task. Spark throws exception only in test when the memory is not released manually before task end.

cc @viirya @maryannxue

**Review by viirya — APPROVED** [review]
Good catch!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 1]
```diff
@@ -695,34 +695,39 @@ case class HashAggregateExec(
         ctx.addInnerClass(generatedMap)
 
         // Inline mutable state since not many aggregation operations in a task
-        fastHashMapTerm = ctx.addMutableState(fastHashMapClassName, "vectorizedHastHashMap",
-          v => s"$v = new $fastHashMapClassName();", forceInline = true)
-        ctx.addMutableState(s"java.util.Iterator<InternalRow>", "vectorizedFastHashMapIter",
+        fastHashMapTerm = ctx.addMutableState(
+          fastHashMapClassName, "vectorizedHastHashMap", forceInline = true)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: While we are touching this, shall we fix this typo `vectorizedHastHashMap` -> `vectorizedFastHashMap`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 2]
```diff
@@ -695,34 +695,39 @@ case class HashAggregateExec(
...
     // Create a name for the iterator from the regular hash map.
     // Inline mutable state since not many aggregation operations in a task
     val iterTerm = ctx.addMutableState(classOf[KVIterator[UnsafeRow, UnsafeRow]].getName,
       "mapIter", forceInline = true)
     // create hashMap
     val hashMapClassName = classOf[UnsafeFixedWidthAggregationMap].getName
-    hashMapTerm = ctx.addMutableState(hashMapClassName, "hashMap",
-      v => s"$v = $thisPlan.createHashMap();", forceInline = true)
+    hashMapTerm = ctx.addMutableState(
+      hashMapClassName, "hashMap", forceInline = true)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: nit. One line?
```scala
hashMapTerm = ctx.addMutableState(hashMapClassName, "hashMap", forceInline = true)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 3]
```diff
@@ -686,7 +686,7 @@ case class HashAggregateExec(
     val thisPlan = ctx.addReferenceObj("plan", this)
 
     // Create a name for the iterator from the fast hash map.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Could you update this comment accordingly in this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 4]
```diff
@@ -686,7 +686,7 @@ case class HashAggregateExec(
...
-    val iterTermForFastHashMap = if (isFastHashMapEnabled) {
+    val (iterTermForFastHashMap, createFastHashMap) = if (isFastHashMapEnabled) {
       // Generates the fast hash map class and creates the fash hash map term.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: nit. typo. `fash hash` -> `fast hash`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 5]
```diff
@@ -695,34 +695,39 @@ case class HashAggregateExec(
         ctx.addInnerClass(generatedMap)
 
         // Inline mutable state since not many aggregation operations in a task
-        fastHashMapTerm = ctx.addMutableState(fastHashMapClassName, "vectorizedHastHashMap",
-          v => s"$v = new $fastHashMapClassName();", forceInline = true)
-        ctx.addMutableState(s"java.util.Iterator<InternalRow>", "vectorizedFastHashMapIter",
+        fastHashMapTerm = ctx.addMutableState(
+          fastHashMapClassName, "vectorizedHastHashMap", forceInline = true)
+        val iter = ctx.addMutableState(
+          s"java.util.Iterator<InternalRow>",
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: `s"` -> `"`

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with only minor comments). Thank you, @cloud-fan .
cc @dbtsai

**Comment by viirya** [comment]
retest this please.

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks for the review, merging to master/2.4!

**Comment by dongjoon-hyun** [comment]
Thank you so much!

**Pull request closed without merging**